Fix print resume with UBL

espr14 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description
Resume print uses current_position that does not account with UBL's Z modifier. This could cause hit of the bed and/or the printed object. This fix uses higher level function that accommodates it.

<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

### Requirements
- UBL
<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits
- works with all variants of bed leveling
<!-- What does this PR fix or improve? -->

### Configurations
#define AUTO_BED_LEVELING_UBL

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues

<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

espr14 committed:
Fix print resume with UBL

Resume print uses current_position that does not account with UBL's Z modifier. This fix uses higher level function that accommodates it.

thinkyhead commented on Marlin/src/module/motion.cpp:
```diff
@@ -1049,6 +1049,17 @@ FORCE_INLINE void segment_idle(millis_t &next_idle_ms) {
+  /// save default value
+  feedRate_t dfr = feedrate_mm_s;
+  destination = current_position;
+  destination.z += z;
```
Since this adds the given value to the current Z position, this function should be called `prepare_z_raise`.

thinkyhead committed:
Use prepare_internal_move_to_destination

thinkyhead committed:
Patch some old axis indexes

thinkyhead commented:
I've modified this PR so it uses the `prepare_internal_move_to_destination` method, because in addition to being compatible with UBL motion, it also does moves without the feedrate multiplier. Following the same idea, I also modified some other code that used the "blocking move" methods. Please confirm that this does all the right things!

espr14 commented on Marlin/src/gcode/feature/pause/M701_M702.cpp:
```diff
@@ -113,8 +117,12 @@ void GcodeSuite::M701() {
-    do_blocking_move_to_z(_MAX(current_position.z - park_point.z, 0), feedRate_t(NOZZLE_PARK_Z_FEEDRATE));
+  if (park_point.z > 0) {
+    destination = current_position;
+    destination.z -= park_point.z;
```
The lift could be smaller due to `NOMORE(destination.z, Z_MAX_POS);`. Z position has to be saved and restored.

espr14 commented on Marlin/src/feature/pause.cpp:
```diff
@@ -597,11 +597,13 @@ void resume_print(const_float_t slow_load_length/*=0*/, const_float_t fast_load_
+    // Move Z back to saved position
+    destination.z = resume_position.z;
+    prepare_internal_move_to_destination(NOZZLE_PARK_Z_FEEDRATE);
   }
```
Previously blocking, now non-blocking move. Could cause an issue. Similar in other places. `planner.sychronize()` missing?

thinkyhead replied:
`planner.synchronize` is only needed if the code is doing something other than motion and therefore needs to wait for previous motion to finish. This is clearer when it is used in front of the non-motion code instead of after the motion code.

espr14 commented:
> I've modified this PR so it uses the `prepare_internal_move_to_destination` method, because in addition to being compatible with UBL motion, it also does moves without the feedrate multiplier. Following the same idea, I also modified some other code that used the "blocking move" methods. Please confirm that this does all the right things!

I don't have a suitable printer to test it. Since there were blocking moves and now you have to set destination as well, it would be better to define similar function e.g. `do_blocking_UBL_move` to keep the code short and readable.

thinkyhead committed:
Pre-apply raise limit

thinkyhead commented:
> do_blocking_UBL_move

Nay, we'd rather not make more special cases for UBL, when we should instead bring UBL into line and integrate it better with the planner.

thinkyhead committed:
save a byte or two

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21564

thinkyhead committed:
limit and lambda
